tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,358 @@
+    ///
+    /// [`NotInitializedError`]: crate::sync::NotInitializedError
+    pub fn get(&self) -> Result<&T, NotInitializedError> {
+        if self.initialized() {
+            Ok(unsafe { self.get_unchecked() })
+        } else {
+            Err(NotInitializedError(()))
+        }
+    }
+
+    /// If the cell is initialized, this method returns a mutable reference to its value,
+    /// otherwise returns [`NotInitializedError`].
+    ///
+    /// [`NotInitializedError`]: crate::sync::NotInitializederror
+    pub fn get_mut(&mut self) -> Result<&mut T, NotInitializedError> {
+        if self.initialized() {
+            Ok(unsafe { self.get_unchecked_mut() })
+        } else {
+            Err(NotInitializedError(()))
+        }
+    }
+
+    /// Sets the value of the OnceCell to the argument value.
+    ///
+    /// If the value of the OnceCell was already set prior to this call
+    /// then [`AlreadyInitializedError`] is returned. If another thread
+    /// is initializing the cell while this method is called,
+    /// ['InitializingError`] is returned. In order to wait
+    /// for an ongoing initialization to finish, call
+    /// [`OnceCell::get_or_init_with`] instead.
+    ///
+    /// [`AlreadyInitializedError`]: crate::sync::AlreadyInitializedError
+    /// [`InitializingError`]: crate::sync::InitializingError
+    /// ['OnceCell::get_or_init_with`]: crate::sync::OnceCell::get_or_init_with
+    pub fn set(&self, value: T) -> Result<(), SetError> {
+        if !self.initialized() {
+            // Another thread might be initializing the cell, in which case `try_acquire` will
+            // return an error
+            match self.semaphore.try_acquire() {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        return Ok(());
+                    } else {
+                        unreachable!(
+                            "acquired the permit after OnceCell value was already initialized."
+                        );
+                    }
+                }
+                _ => {
+                    // Couldn't acquire the permit, look if initializing process is already completed
+                    if !self.initialized() {
+                        return Err(SetError::InitializingError(()));
+                    }
+                }
+            }
+        }
+
+        Err(SetError::AlreadyInitializedError(()))
+    }
+
+    /// Tries to set the value of the cell, overwriting the previously set value, in case one is
+    /// available. If no value was previously set, this method has the same functionality has
+    /// [`OnceCell::set`].
+    ///
+    /// [`OnceCell::set`]: crate::sync::OnceCell::set
+    pub fn set_mut(&mut self, value: T) -> Result<(), SetError> {
```

Comment:
Design question: This method can unconditionally set the value, overwriting the previous one if one exists. Which one is best?

Reply:
I'd prefer not being able to set the value if a previous one exists. To me this violates the purpose of `OnceCell`. Even if you have access to a mutable reference, it makes no sense to have that functionality in my opinion.

Reply:
Then just delete this function. There's no real reason to have it if not for this feature.